### ISSUE TITLE
Sieges: bugs & missing features

### ISSUE THREAD

**Issue by IvanSavenko** [vcmi/vcmi#1236]
Split from #1219 for convenience, planning to do separately due to its relatively large size.
If somebody is willing to do testing for me (all "Check X" tasks) using current develop branch that would be appreciated.

Misc:
- [ ] Check Teleport spell (only Advanced+ should allow teleporting inside walls)
- [ ] Check whether Cyclops wall attacks are functional
- [ ] Check whether Cyclop attack matches Basic Ballistics params (or Advanced for Cyclop Kings)

Castle Gates: https://heroes.thelazy.net/index.php/Wall
- [ ] castle gates should render unused now image ****DrwC.bmp with front part of gate (chains). Needed to correctly render stacks standing on top of the gate's bridge
- [ ] castle gates have no animation to lower/raise bridge
- [ ] if catapult makes multiple shots, gates will be destroyed only after final shot
- [ ] Check whether placing unit on tile to the right of the drawbridge will open it
- [ ] Check whether placing Force Field on moat will stop it from opening (that's a thing, apparently)

Castle Walls: https://heroes.thelazy.net/index.php/Wall
- [ ] Walls should change visual to more damaged state at midpoint of explosion playback instead of after it
- [ ] Walls should get extra hitpoint once Castle is built
- [ ] Check whether walls correctly incur ranged attacks penalty (and correctly update after damage to walls)

Arrow Towers: https://heroes.thelazy.net/index.php/Arrow_Tower
- [ ] Stack from destroyed tower is removed instantly, should be done at midpoint of the explosion
- [ ] Check whether Arrow Towers receive damage bonus depending on number of built structures in town

Moat: https://heroes.thelazy.net/index.php/Moat
- [ ] Check whether creatures in moat receive additional 10% damage from physical attacks
- [ ] Check whether creatures in moat have their defense reduces by 3
- [ ] Check whether Fortress moat is 2-tiles wide
- [ ] Check whether Tower moat (Mines) deals correct damage
- [ ] Check whether Inferno & Fortress moat deals 90 damage
- [ ] Check whether all other towns' moat deals 70 damage

Ballistics: https://heroes.thelazy.net/index.php/Ballistics
- [ ] Implement weighted-chance target selection 
- [ ] Make each attack into independent Pack, for easier animation ordering
- [ ] Check whether probability to hit intended wall matches H3
- [ ] Check whether dealt damage matches H3

**Comment by rilian-la-te** [comment]
@IvanSavenko is Arrow Towers not ignoring defence now?

**Comment by IvanSavenko** [comment]
Correct. They have IGNORE_DEFENSE bonus or what's its name. Updated list.

**Comment by rilian-la-te** [comment]
@IvanSavenko please, include moats here too.
